src/x/cache/lru_cache.go

```diff
@@ -430,10 +430,15 @@ func (c *LRU) handleCacheLoadErrorWithLock(
 // updateCacheEntryWithLock updates a cache entry with a new value or cached error,
 // and marks it as the most recently accessed and most recently loaded entry
 func (c *LRU) updateCacheEntryWithLock(
-	key string, expiresAt time.Time, value interface{}, err error,
+	key string, expiresAt time.Time, value interface{}, err error, enforceLimit bool,
 ) (interface{}, error) {
 	entry := c.entries[key]
 	if entry == nil {
+		if enforceLimit && c.reserveCapacity(1) != nil {
+			// Silently skip adding the new entry if we fail to free up space for it
+			// (which should never be happening).
+			return value, err
+		}
```

Comment:
Should return the error from the reserve capacity call yeah? Otherwise the err would be nil yeah?
```suggestion
		if enforceLimit {
			if err := c.reserveCapacity(1); err != nil {
				// Silently skip adding the new entry if we fail to free up space for it
				// (which should never be happening).
				return value, err
			}
		}
		
```

Reply:
Well, this aspect is a bit vague. The contract of `updateCacheEntryWithLock` is to return its arguments `value interface{}, err error` unmodified, and I've tried to preserve that, commenting that we "silently skip" such situation (being unable to evict from cache - which from what I see would only happen in case there is a code bug).
Also, in the only place that calls `updateCacheEntryWithLock` with `enforceLimit = true`, those return values are ignored completely:
https://github.com/m3db/m3/blob/c5f6237fd98b03fb808f159e644ad6a9264e9dca/src/x/cache/lru_cache.go#L223-L233

So I'm not really sure if I should change this (but don't have a strong opinion either, even after jumping around this code for quire a long time). I guess that's the support of two modes (with/without loader) in a single data structure that makes it hard to come up with an elegant implementation. Perhaps @ryanhall07 will chime in with his insights as well.

Reply:
I see yeah, that makes sense.

The other question is, this only happens when there is a code bug yeah? 

If so.. does that mean we have a remaining code bug we're not aware of? Which is fine if so, we can chase this up after the fact. Just wanted to understand current state of the world (post-merging this change).

Reply:
No, what I meant, `reserveCapacity` is not supposed to return errors, unless there is some bug that we are not yet aware of.

Reply:
Ah ok, that makes sense.

Reply:
left another comment. maybe it's just best to change `reserveCapacity` to return a bool so it's clear it can't be an error.